Client Side Development

anhnamtran opened the pull request:
An implementation for the client side of the Tellina user experiment infrastructure.

All files added are in the `client_side/` directory.

Higher priority files:
- `configure`: sources `.infrastructure/setup.sh`
- `.infrastructure/setup.sh`: constants and initial value definitions.
- `.infrastructure/infrastructure.sh`: function definitions for
  infrastructure to run smoothly.
- `.infrastructure/verify_task.py`: primary logic for task output
  verification.

Lower priority files:
- `.infrastructure/tasks/`: directory with all tasks descriptions and expected outputs.
- `.infrastructure/bash-preexec.sh`: downloaded from
  [here](https://github.com/rcaloras/bash-preexec) using their provided
  instructions. Script that installs Bash Preexec.
- `fs.tgz`: tarball with file system directory for users to work on.

anhnamtran committed:
Set up directory for tasks

anhnamtran committed:
Added some preliminary infrastructure code

anhnamtran committed:
Added more to infrastrcture code

anhnamtran committed:
Print error for get task description

anhnamtran committed:
Fixed some bugs relating to working directories

anhnamtran committed:
Updated output verification

anhnamtran committed:
Change where USER_OUT is defined

anhnamtran committed:
Changed location of an echo

anhnamtran committed:
Fixed a silly typo

anhnamtran committed:
Make sure that infrastructure scripts are executable

anhnamtran committed:
Detect if user has no display for diff

anhnamtran committed:
Remove trailing whitespaces

anhnamtran committed:
No need to close files

anhnamtran committed:
Added another check

anhnamtran committed:
Reset now clears the interface

anhnamtran committed:
Correctly handles case when experiment first starts

anhnamtran committed:
Use the correct operator

anhnamtran committed:
Moved timeout message

anhnamtran committed:
Merge branch 'master' into client-dev

anhnamtran committed:
Merge branch 'master' into client-dev

anhnamtran committed:
Minor redesigns

- Implemented determine task order differently
- Some style changes

anhnamtran committed:
Merge branch 'master' into client-dev

anhnamtran committed:
Renamed task files

anhnamtran committed:
Changed log header task_no -> task_code

anhnamtran committed:
Removed a column from log

anhnamtran committed:
Removed unnecessary line from expected output

anhnamtran committed:
Added file for bash-preexec installation

anhnamtran committed:
Update gitignore

anhnamtran committed:
Correctly gets full path for configure script

anhnamtran committed:
Got rid of two unneeded scripts

anhnamtran committed:
Implementation for setup.sh

anhnamtran committed:
Implementation for infrastructure.sh

anhnamtran committed:
Changed verify_task.py to adhere to design doc

anhnamtran committed:
Fixed bug with char_from

anhnamtran committed:
Added treatment and task set changing in next task

anhnamtran committed:
Now "uninstalls" Bash pre-exec on experiment end

anhnamtran committed:
Moved a few lines of code around

anhnamtran committed:
Better printing for console

anhnamtran committed:
Small typo

anhnamtran committed:
Added some documentation

anhnamtran committed:
Added ability to resume experiment

anhnamtran committed:
More documentation

anhnamtran committed:
Formatting

anhnamtran committed:
Edited README file

mernst committed:
Simplify comment

mernst committed:
Tweaks from code review

mernst commented on client_side/.infrastructure/setup.sh:
```diff
@@ -0,0 +1,238 @@
+################################################################################
+
+# Saves the old value of PROMPT_COMMAND, since Bash Preexec overwrites it.
+PROMPT_COMMAND_OG=${PROMPT_COMMAND}
```
I find the suffix "_OG" confusing.  I would use "_ORIG" or "_USER" instead.

anhnamtran replied:
Resolved in 5381c5b39de29628a4ae8476272c13973d0fd451.

mernst commented on client_side/.infrastructure/setup.sh:
```diff
@@ -0,0 +1,238 @@
+# The absolute path to the user experiment directory
+EXP_DIR="$1"
+# The absolute path to the experiment's infrastructure directory
+INFRA_DIR="${EXP_DIR}/$(dirname ${BASH_SOURCE[0]})"
```
It is possible that `${BASH_SOURCE[0]}` is already an absolute filename, in which case INFRA_DIR will be an invalid directory.

anhnamtran replied:
`$BASH_SOURCE` contains [the relative path](https://stackoverflow.com/a/35006505) of the containing script. For example,
```sh
$ ls -R
.:
script1 script2 dir

./dir:
script3
$ cat script1
#!/bin/bash
source ./dir/script3
$ cat script2
#!/bin/bash
cd ./dir
source ./script3
$ cat dir/script3
#!/bin/bash
echo "${BASH_SOURCE[0]}"
$ ./script1
./dir/script3
$ ./script2
./script3
```

Because `configure` [`cd`'s](https://github.com/anhnamtran/tellina_user_experiment/pull/3/files#diff-e5ddb23605f4c5338bc03c469733d118R6) into `${EXP_DIR}` then runs
```sh
source .infrastructure/infrastructure.sh
```
The `${BASH_SOURCE[0]}` will always return `.infrastructure/infrastructure.sh`.

mernst commented on client_side/.infrastructure/setup.sh:
```diff
@@ -0,0 +1,238 @@
+INFRA_DIR="${EXP_DIR}/$(dirname ${BASH_SOURCE[0]})"
+
+# Establish tasks directories and related variables
+TASKS_SIZE=22
```
I suggest you compute this by doing `ls` of the directory, rather than hard-coding it.

anhnamtran replied:
Resolved in 5381c5b39de29628a4ae8476272c13973d0fd451.

anhnamtran replied:
Resolved in 8af6cc0.

mernst commented on client_side/.infrastructure/setup.sh:
```diff
@@ -0,0 +1,238 @@
+TASKS_SIZE=22
+
+TASKS_DIR="${INFRA_DIR}/tasks"
+TIME_LIMIT=300
```
This is a *per-task* time limit, and I would change the name to reflect that.

anhnamtran replied:
Resolved in 5381c5b39de29628a4ae8476272c13973d0fd451.

mernst commented on client_side/.infrastructure/setup.sh:
```diff
@@ -0,0 +1,238 @@
+SERVER_ROUTE="/tellina_user_experiment/server_side/post_handler/post_handler.php"
+
+# Establish survey URL
+SURVEY_URL="<URL>"
```
I suggest that you put the survey in this same repository, in which case the URL can also be under SERVER_HOST.  Go ahead and create the HTML file now and set the survey URL.

anhnamtran replied:
Partial resolution in 5381c5b39de29628a4ae8476272c13973d0fd451.

The `SURVEY_URL` variable is now `EXPERIMENT_HOME_URL` that links to the [home page](https://homes.cs.washington.edu/~atran35/tellina_user_experiment) of the experiment. I'm assuming that we will post the survey link there.

mernst commented on client_side/.infrastructure/setup.sh:
```diff
@@ -0,0 +1,238 @@
+MACHINE_NAME=$(hostname)
+read -p "Enter your UW NetID: " USER_NAME
+
+USER_ID="${USER_NAME}@${MACHINE_NAME}"
```
I said before that I think the machine name should not be used.  It's extra for the user to remember to type later.
It would only be necessary if we were concerned that two students would have identical usernames (each on their own private VM).  But if students are entering their UW NetID, then there will be no collisions and the machine name is not necessary.

anhnamtran replied:
The machine name is only used for the log file, as in `$USER_ID` is used in write log. `$USER_NAME` is used every else.

anhnamtran replied:
Would you like the log file to not have machine name either?

mernst replied:
If the log file contains the machine name, I suggest putting it in a separate column to simplify analysis.

anhnamtran replied:
Considering that the `user_id` column always separates the `user_name` and
the `machine_name` with an `@` symbol, would adding another column to the log file be any different than splitting the text in analysis?

mernst replied:
The only difference is less work to do in the analysis phase.  Every time we write code -- however simple -- that increases the burden of understanding the code and increases the likelihood of an error.

mernst commented on client_side/.infrastructure/setup.sh:
```diff
@@ -0,0 +1,238 @@
+#             This includes bash variables as well as variable files           #
+################################################################################
+
+# Makes sure that all the scripts are executable
```
I think it would be cleaner to ensure they are right in the zip file, and then this is not needed.
Where is the Makefile that creates the distribution zip file?  Please create it and check it into this repository.

anhnamtran replied:
The Makefile has just been added by commit a756d7839. Please take a look. I went ahead and made it so the file runs tests first before creating the distribution files.

mernst commented on client_side/.infrastructure/setup.sh:
```diff
@@ -0,0 +1,238 @@
+case $(echo $((0x$(md5sum <<<${USER_NAME} | cut -c1) % 4))) in
+  0)
+    echo "T1N2" > "${INFRA_DIR}/.task_order"
+    echo "T" > "${INFRA_DIR}/.treatment"
```
I think this case statement should only set .task_order.  I think there should be a utility function that sets .treatment and task_set from .task_order and curr_task.  That will reduce code duplication here and elsewhere.

mernst commented on client_side/.infrastructure/infrastructure.sh:
```diff
@@ -0,0 +1,201 @@
+  make_fs
+
+  # Increment the number of tasks finished by the user
+  curr_task=$(( curr_task + 1 ))
```
The variable name doesn't indicate whether it is a task number, task code, or something else.  I would call it task_num instead.

anhnamtran replied:
Resolved in 5381c5b.

mernst replied:
You don't have to respond to every comment individually.  If you change the line the comment is on, I believe GitHub will automatically not show the comments in the "Files changed" tab.

mernst commented on client_side/.infrastructure/setup.sh:
```diff
@@ -0,0 +1,238 @@
+# Executed after the user-entered command is executed
+# - Only one of these cases can happen
+# 1. Check if user has run out of time:
+#    - `time_elapsed=$SECONDS` is less than some time limit constant.
```
I think this description is inverted.

anhnamtran replied:
Resolved in 8af6cc0.

mernst commented on client_side/.infrastructure/setup.sh:
```diff
@@ -0,0 +1,238 @@
+#      `time_elapsed=$TIME_LIMIT`.
+# 2. Handle user meta-command:
+#    - Output verification will not be performed on these commands.
+#    - The check is done by looking for the existence of the file
```
I think it would be cleaner to put the functionality in the aliases for `abandon`, etc instead.  That would dispense with the need to create and delete these marker files.
You would still have a marker file for .noverify (meaning that no output verification should be done, which is relevant to all 4 special commands and to starting the task), but this is simpler than having 5 `if`s in this file.

anhnamtran replied:
Commit 8af6cc0 has a solution that is up for consideration.

mernst commented on client_side/.infrastructure/setup.sh:
```diff
@@ -0,0 +1,238 @@
+#      - If `helpme`: print the list of user meta-commands. Remove `.helpme`.
+#      - If `task`: call `get_task_description.py` with `.task_code` to print the
+#        task's description. Remove `.task`
+#      - Set `status="incomplete"`.
```
Why does it have to set this?  Isn't status guaranteed to be incomplete if execution gets to here?  If that isn't the case, it might be nice to document the possibilities.

anhnamtran replied:
It does not.

mernst commented on client_side/.infrastructure/setup.sh:
```diff
@@ -0,0 +1,238 @@
+
+      meld "/tmp/task_actual" "/tmp/task_expected" &
+    elif [[ -z ${status} ]]; then
+      status="Verification error!"
```
Say what to do, such as how to report it or where to find diagnostic output.

mernst commented on client_side/.infrastructure/setup.sh:
```diff
@@ -0,0 +1,238 @@
+make_fs
+cd "${FS_DIR}"
+
+# Prints the introduction here
```
I feel this belongs in start_experiment, not in `setup.sh`.

mernst commented on client_side/.infrastructure/infrastructure.sh:
```diff
@@ -0,0 +1,201 @@
+
+# Converts a numeric ASCII value to a character.
+chr() {
+  [ "$1" -lt 256 ] || return 1
```
I think it would be better to crash here (certainly during testing!), rather than silently returning an incorrect value.

anhnamtran replied:
The actual "return" value of `chr` is what it prints not the `return` statement. `return` here is dictating its return code. There are now comments documenting this behavior for both `chr` and `ord`.

mernst commented on client_side/.infrastructure/infrastructure.sh:
```diff
@@ -0,0 +1,201 @@
+}
+
+# Outputs the nth alphabetic character. n is 1-based; that is, char_from(1) is "a".
+# $1: the number n specifying which characters
```
Why is "characters" plural?

anhnamtran replied:
Resolved in 8af6cc0.

mernst commented on client_side/.infrastructure/infrastructure.sh:
```diff
@@ -0,0 +1,201 @@
+# The user task number is always sequential.
+# The task can be in either task set 1 or 2, with the dividing task being
+# TASKS_SIZE / 2.
+# That is, if the current task set is 2 and the user current user task number
```
This seems inconsistent with the code.  We should be in the else branch because task_set is 2. In the else branch, no subtraction is done so this cannot return a value less than the current user task number.

mernst commented on client_side/.infrastructure/infrastructure.sh:
```diff
@@ -0,0 +1,201 @@
+# That is, if the current task set is 2 and the user current user task number
+# is 12, then its true task code is "a", because it is in task set 1.
+get_task_code() {
+  if ((task_set == 1)); then
```
I found this logic hard to follow.

One possibly-simpler way to think about this is that there are these cases:
 * Second character of treatment = "1":  nothing to do
 * Second character of treatment = "2":  need to add or remove TASKS_SIZE/2.

Or, maybe better documentation of the current logic.

Also, please write test cases for the functions in this file, such as chr, ord, char_from, get_task_code.  Put those test cases in a separate file.

anhnamtran replied:
Commit f5845dd4 has added a `test` directory under `.infrastructure` which uses Bats to perform the tests. I would really like to hear your thoughts on this.

Over the weekend I will add several spec tests to ensure that all infrastructure code works as well (`next_task`, `precmd`, `preexec`, etc.). I believe the spec tests will help us in the long run as well.

With the addition of the tests, I found out that `get_task_code` does not handle an odd number of tasks (i.e. `$TASKS_SIZE` is odd) very well. The two tests `get_task_code correct - ... odd task size` are skipped because of this reason.

mernst replied:
That seems fine.

One concern is that this requires another package, so anyone developing on it needs to install bats.  The dependency must not leak out to experimental subjects.  It probably won't, but it's just one more concern.

Another concern with f5845dd is that it's hard to figure out where the tests are and what they do, and they depend on new code written in the tests.  I had expected about  two lines of test for each of the bash functions.  If the function works for 'a' and 'q', then it's hard to imagine that it would not work for all inputs, and it's easy to manually validate the tests and better understand the function's operation if it was not clear from the documentation.  These tests don't do either of those things.  So, I feel that they are overkill.

That said, it's great that adding the tests revealed a limitation.  Maybe the Makefile task that creates the zip file should fail if there are an odd number of tasks.

mernst commented on client_side/.infrastructure/infrastructure.sh:
```diff
@@ -0,0 +1,201 @@
+  precmd_functions+=(precmd_func)
+
+  print_treatment
+  print_task
```
The next three lines should be abstracted out into a start_task function.

mernst commented on client_side/.infrastructure/infrastructure.sh:
```diff
@@ -0,0 +1,201 @@
+  preexec_functions+=(preexec_func)
+  precmd_functions+=(precmd_func)
+
+  print_treatment
```
There should perhaps be a start_treatment function that also does training if necessary.

mernst commented on client_side/.infrastructure/infrastructure.sh:
```diff
@@ -0,0 +1,201 @@
+
+  rm -rf "${FS_DIR}"
+  mkdir "${FS_DIR}" > /dev/null
+  tar -xzf "${INFRA_DIR}/fs.tgz" -C "${FS_DIR}"
```
I am concerned that this command produces output and you have not supplied the `q` command-line option.

Also, I think that `rsync` is a better choice for the implementation than `rm` followed by `tar`.  If you do that, you don't even need to save and restore the user's directory.

anhnamtran replied:
Does this mean that instead of using `fs.tgz` it would just be another `file_system` directory under `.infrastructure` that is the original version that can be used for `rsync`?

Or should it be keep the tarball, extract it only once, then use `rsync` on the extracted directory?

mernst replied:
Either of those approaches would be fine.

mernst commented on client_side/.infrastructure/infrastructure.sh:
```diff
@@ -0,0 +1,201 @@
+  if [[ "$(cat "${INFRA_DIR}/.treatment")" == "T" ]]; then
+    echo "and Tellina <URL> to help you solve the tasks."
+  else
+    echo "but you can't use Tellina to help you solve the tasks."
```
If the first treatment is N (NoTellina), then the user should not have even been introduced to Tellina yet, so this does not need to be printed.  In that case, training on Tellina should occur after the halfway mark.  So, there are two types of training:
 * on the infrastructure
 * on Tellina
and users may receive both at the very beginning of the experiment, or one at the beginning of each half of the experiment.

anhnamtran replied:
Resolved in 8af6cc0.

mernst commented on client_side/.infrastructure/infrastructure.sh:
```diff
@@ -0,0 +1,201 @@
+  local task_code="$(cat "${INFRA_DIR}/.task_code")"
+  local user_command="$(cat "${INFRA_DIR}/.command")"
+
+  status=$("${INFRA_DIR}"/verify_task.py ${task_code} ${user_command})
```
I am concerned about the lack of quoting of `${user_command}` on this line.
Quoting commands correctly to pass them to another program is a bit of a mess.  Should `verify_task.py` take a filename instead of (or as an alternate to) the command itself?

anhnamtran replied:
The current implementation for `verify_task.py` is that every argument after the first (i.e. 2, 3, ...) is part of the user's command. With that said, I don't see a case where that would be a problem. 

If the user command is properly quoted and escaped, so should that passed to `verify_task.py`. If it is not, then there would be an error in the shell, and `verify_task.py` should fail as well.

mernst replied:
I know I've been bitten by this in the past.  I have attached [a small test case](https://github.com/anhnamtran/tellina_user_experiment/files/3148745/bash_test.zip).  It should print 3 lines, but instead it prints 4.  
Here is the result of executing it:

```
% bash_test.sh
echo
bbb
"ccc
ddd"
```

Are you sure that your code will not be susceptible to a problem like this?
You might want to store each argument on its own line in .command, unquoted.

mernst commented on client_side/.infrastructure/infrastructure.sh:
```diff
@@ -0,0 +1,201 @@
+    print_treatment
+  fi
+
+  echo $(get_task_code) > "${INFRA_DIR}/.task_code"
```
From here down should be abstracted out into a start_task function.
Maybe it should also incorporate most of the logic above and handle switching the treatment too (or calling a separate function which does so).  Note that switching the treatment may require more training.

mernst commented on client_side/.infrastructure/infrastructure.sh:
```diff
@@ -0,0 +1,201 @@
+    -d treatment="$(cat "${INFRA_DIR}/.treatment")" \
+    -d time_elapsed="${time_elapsed}" \
+    -d status="${status}" \
+    -d command="$(cat "${INFRA_DIR}/.command")" &> /dev/null
```
Is there any chance this will fail?
If so, I suggest putting the output in a file, and if the status was failure, printing it -- at least during development and pilot studies.
If not, this is fine as is.

anhnamtran replied:
This will fail under the following conditions:
1. If the user is not connected to the internet.
2. If the server is somehow down. This would require my CSE Homes to be down.
3. Internal server error. That is, the POST handler is buggy.
4. The `$SERVER_HOST` and/or `$SERVER_ROUTE` variables are incorrect.

mernst reviewed (changes requested):
Thanks for this!  It is looking good.  I have some comments.  Let me know if you have questions.

Minor writing issues:
Where you wrote "full path", the technical term is "absolute path".
Please end sentences with a period.  If the period is missing, the reader has to halt and think about whether the sentence was unintentionally truncated, or the author just didn't bother to write the period.

mernst committed:
Simplify file names

mernst committed:
Code review improvements

mernst commented on client_side/.infrastructure/verify_task.py:
```diff
@@ -0,0 +1,183 @@
+        def __exit__(self, etype, value, traceback):
+            os.chdir(self.savedPath)
+
+    # the current task number, as a str
```
This looks like the true task code

anhnamtran replied:
Yes, this comment slipped through the cracks.

anhnamtran replied:
Resolved in 8af6cc06a5d30b1f6a4ab6d5ef43f66d6cbe83df.

anhnamtran committed:
Merge branch 'master' into client-dev

anhnamtran committed:
Merge branch 'master' into client-dev

anhnamtran committed:
Merge branch 'master' into client-dev

anhnamtran committed:
Renamed several variables

anhnamtran committed:
Merge branch 'master' into client-dev

anhnamtran committed:
Addressed some code review suggestions

- TASK_SIZE is now computed instead of a magic number
- User meta-command aliases now invoke the functions themselves instead
  of relying on precmd.
  - They create a file called `.noverify` that precmd checks to skip
    task verification.
- precmd no longer checks for separate `.<command_name>` files, instead
  it just checks for `.noverify` and deletes it afterwards
- Moved a couple of lines into start experiment.

anhnamtran committed:
Changed an output and some comments

anhnamtran committed:
Added several test files for utility code

anhnamtran committed:
Added TODO to get_task_code

anhnamtran committed:
Added extra documentation to chr and ord

anhnamtran committed:
Added more output printing for status assertions

anhnamtran committed:
Added an extra check and test for chr

anhnamtran committed:
Added a top level comment for utility tests

anhnamtran committed:
Removed test name printing from utility functions tests

anhnamtran committed:
Added an assertion function

anhnamtran committed:
Added several specification tests for make_fs

anhnamtran committed:
Removed chmod from setup.sh

anhnamtran committed:
Added Makefile for tests and distributions

anhnamtran committed:
Test target now prints machine parseable output

anhnamtran committed:
Added .travis.yml file

anhnamtran committed:
Changed Makefile

anhnamtran committed:
Changed how assert_\(status\) works

anhnamtran committed:
Travis runs make instead

anhnamtran committed:
Temporarily disabled email notifications during setup

anhnamtran committed:
Install Python 3 on travis

anhnamtran committed:
Removed language tag from Travis

anhnamtran committed:
Use subprocess.call instead for compability with < python3.5

anhnamtran committed:
[skip travis] Made machine name a separate column

anhnamtran committed:
Re-enable email on failure

anhnamtran committed:
Merge branch 'master' of github.com:anhnamtran/tellina_user_experiment into client-dev

anhnamtran committed:
Removed TODO for handling odd task codes

anhnamtran committed:
Use correct variable for link to post handler

anhnamtran committed:
Update gitignore

anhnamtran committed:
Merge branch 'master' of github.com:anhnamtran/tellina_user_experiment into client-dev

anhnamtran committed:
Refactored code into separate functions

anhnamtran committed:
Greatly simplified utility functions tests

anhnamtran committed:
Added infrastructure tests file

anhnamtran committed:
Changed design doc to match code changes during review

anhnamtran committed:
Made the experiment half number more intuitve

anhnamtran committed:
Cleaned up test directory and made things more organized

anhnamtran committed:
Added a Makefile in client_side/.infrastructure and updated root Makefile

- `make test` can now be run in both the `client_side/.infrastructure`
  directory and the root directory.
- `make distribution` now depends on `client_side` files instead of just
  the direcotry.
- The distributed ZIP file now stores the client side directory as the
  distribution name instead of the `client_side` name.

anhnamtran commented:
@mernst Things that still need to be added:
- Tests:
  - Passing user command with quotes to `verify_tasks.py`.
  - More specification tests for the infrastructure: `next_task`, `precmd_func`, `preexec_func`, etc.
    - There will be failure tests for each function as well. So I'm unsure if we should add special error handling and logging for the infrastructure itself.

- READMEs for each `test` subdirectory.
- Training for infrastructure and Tellina (this can possibly be done after we merge this into master).

anhnamtran committed:
Fix printout bug in assert_output

anhnamtran committed:
Moved similar functions into the same test file

anhnamtran committed:
Fixed annother printout bug in assertions

anhnamtran committed:
Some more tests for functions that handle tasks

anhnamtran committed:
Use the correct return code when experiment is ending

anhnamtran committed:
Added more tests for task handling

anhnamtran committed:
[skip travis] Removed redundant --tap flag

anhnamtran committed:
Made debug printout better

anhnamtran committed:
Added helper function for testing verify_task

anhnamtran committed:
Added first test for verify_task

anhnamtran committed:
Use correct function name for teardown

anhnamtran committed:
Fixed teardown code

anhnamtran committed:
More debugging info for verify task test

anhnamtran committed:
Changed the quoting test a bit

anhnamtran committed:
Removed training detection tests

anhnamtran committed:
Changed a test to be more lenient

anhnamtran committed:
Added jq binary to remove dependency on host having it installed

anhnamtran committed:
Updated documentation based on code review

anhnamtran committed:
Changed how user directory is reset

- Renamed `make_fs` to `reset_fs`.
- `reset_fs` uses `rsync` to reset the user's file system directory.
- Added support for `.infrastructure/file_system` directory, which is
  extracted from `fs.tgz` on setup and used throughout experiment to
  reset user directory.
- Changed test setup to accomodate `reset_fs` reliance on the
  infrastructure sync directory.

anhnamtran committed:
Add executable permission to jq binary

anhnamtran committed:
Removed unneeded recovery files

anhnamtran committed:
Use next_task instead of start_task in start_experiment

anhnamtran committed:
Changed how experiment prompts are done

anhnamtran committed:
Moved some code around

anhnamtran committed:
Curl is now silent

anhnamtran committed:
A period

anhnamtran committed:
Changed a variable to be more informative.

anhnamtran committed:
Periods

anhnamtran committed:
Changed helpme alias to adhere to new prompting

anhnamtran committed:
Better comments

anhnamtran committed:
Changed design doc to have the same variable names

anhnamtran committed:
Changed how verify_task works

- `verify_output.py` no longer prints to stdout.
- `verify_task` uses the exit code of `verify_output.py` to determine
  the status and whether to open meld.
- Documentation has been updated to accommodate these changes.
- Tests have been updated to accommodate these changes.

anhnamtran committed:
Line break

anhnamtran committed:
Training prompts

anhnamtran committed:
[skip travis] Took care of TODOs

anhnamtran committed:
Merge branch 'master' into client-dev

anhnamtran commented:
@mernst I believe I have address most of the things you pointed out in your hand-written code review. Over the weekend I will look to get the website and tutorials set up.

anhnamtran committed:
Moved opening meld to precmd from verify_task

anhnamtran committed:
Adjust Makefile to be cleaner and added more documentation

anhnamtran committed:
Updated README

anhnamtran committed:
Added a check for meld in setup.sh

anhnamtran committed:
Removed set_task_set as it was unneeded (just like most light-themed IDEs)

anhnamtran committed:
Move reset_fs from next_task to start_task

anhnamtran committed:
Make meld start in background

anhnamtran committed:
Addressed multiple code review comments

- Documentation improvements.
- Method simplification.
- Formatting changes.

anhnamtran committed:
Better style for Makefile

anhnamtran committed:
Better comments on some functions

anhnamtran committed:
Added several tests for verify_task

 Please enter the commit message for your changes. Lines starting

anhnamtran committed:
Fixed quoting in assertion

anhnamtran committed:
Added safeguard againsts "cd"

anhnamtran committed:
Change how Meld is opened

anhnamtran committed:
[skip travis] Some changes to website files

anhnamtran committed:
Client's README only contains URL to website

anhnamtran committed:
Added documentation for why shell=True is needed for user commands

anhnamtran committed:
Adjusted some targets and dependencies to deal with symlinks

anhnamtran committed:
Added targets to publish or stage distribution file to host

anhnamtran committed:
Corrected a task code in the list of file system tasks

anhnamtran committed:
Better handling of Meld

anhnamtran committed:
Added tests for cd and pipe in commands

anhnamtran committed:
Updated a couple of task descriptions to fit with current interface

anhnamtran committed:
Revert changes to how experiment prompts are printed

anhnamtran committed:
Implemented training

anhnamtran commented:
@mernst 
## Changes:
1. Training is now implemented. I would like you to specifically go through the changes made in commit 53bc1f0. 
2. Documentation has been significantly improved.
3. There is now a make target for upload the distribution, both to the public site and a staging site.
4. There are now tests for quoting, `cd`, and `|` pipes for the infrastructure.
5. I've decided that this branch should be reserved for only the code and not the website. Thus, website content has been moved to the branch `website`.

## TODOs:
1. Web instructions for infrastructure.
2. Web instructions for Tellina.
3. Documentation on how to host the content for the experiment.

I believe that all of the above mentioned TODO items should not hinder the ability to merge this branch to master.

anhnamtran committed:
Added ability to log suppressed outputs

anhnamtran committed:
Use the new hosting URL

anhnamtran committed:
start_task now changes the user to the starting file_system directory

anhnamtran committed:
Changed how the infrastructure handles `cd` command from the user

mernst commented on client_side/.infrastructure/prompts/ending.txt:
```diff
@@ -0,0 +1,2 @@
+Congratulations! You have completed the interactive portion of the, experiment,
```
I think it would be best to integrate these into the main scripts (even if that causes lines longer than 80 characters long).

mernst committed:
Remove prompts/ directory, which is no longer used (#5)

* Typo fix

* Remove prompts/ directory, which is no longer used

anhnamtran committed:
Changed a couple of tasks to adhere to current infrastructure

anhnamtran commented:
I attempted to change the permissions for all the files in the `file_system` directory in the Makefile target `$(FS_DIR)` using `chmod 666 -R $(FS_DIR)`.

However, this prevent me from accessing the directory from my local machine as well as on Attu where I tested it. 

I'm not sure if this is a mistake on my part or if doing it recursively screws it up. 

Should I only change the permissions for regular files and not directories? Since I know directory permissions could be finicky.

mernst commented on client_side/.infrastructure/verify_task.py:
```diff
@@ -0,0 +1,188 @@
+import filecmp
+import tarfile
+
+# Gets all the environment variables and creates the user output directory if it
```
Cut the second part of this comment after changing the implementation as noted below.

mernst commented on client_side/.infrastructure/verify_task.py:
```diff
@@ -0,0 +1,188 @@
+FS_DIR = os.environ['FS_DIR']
+
+USER_OUT_DIR = os.environ['USER_OUT']
+if not os.path.exists(USER_OUT_DIR):
```
I think setup should do this once an for all rather than verify.py doing it every time.

mernst commented on client_side/.infrastructure/verify_task.py:
```diff
@@ -0,0 +1,188 @@
+
+        if not fs_good:
+            if task_code in FILESYSTEM_TASKS:
+                sys.exit(1)
```
I would prefer using constant variables, rather than these hard-coded literals.

mernst commented on client_side/.infrastructure/verify_task.py:
```diff
@@ -0,0 +1,188 @@
+                        # 1. It preserves quoting for user commands given that
+                        # the command argument passed to the call is a string.
+                        # (see https://docs.python.org/3/library/subprocess.html#popen-constructor starting at "On POSIX with shell=True")
+                        # 2. It allows reduces the need to replace shell
```
Typo "allows reduces"

mernst commented on client_side/.infrastructure/verify_task.py:
```diff
@@ -0,0 +1,188 @@
+        print(e)
+        sys.exit(5)
+
+def normalize_output(out_file, norm_file):
```
I found the method name confusing because I thought the normalization might be in place.  I don't mind this design, but how about changing the method name to something like "normalize_and_copy"?

mernst commented on client_side/.infrastructure/verify_task.py:
```diff
@@ -0,0 +1,188 @@
+    Normalizes the contents of file out_file (sorts lines, removes leading './')
+    and writes the result to file norm_file.
+    """
+    norm_out = open(norm_file, 'w')
```
Why not use `with open...` so you don't need the close() statements at the end of the method?

mernst commented on client_side/.infrastructure/verify_task.py:
```diff
@@ -0,0 +1,188 @@
+        if line == './' or line == '.':
+            p_line = line
+        else:
+            p_line = line.lstrip('./')
```
Changing the lines could make the lines no longer be sorted.  (The order will be deterministic, just not sorted.)
I suggest first modiying the lines, then sorting, then printing.  This requires two loops instead of one, but it's worth it.

mernst commented on client_side/.infrastructure/verify_task.py:
```diff
@@ -0,0 +1,188 @@
+
+
+def verify(norm_file, task_code, check_fs):
+    """Returns 0 if verification succeeded, non-zero if it failed."""
```
This method returns a boolean, not a number.

mernst committed:
Initial code review changes

mernst commented on client_side/.infrastructure/setup.sh:
```diff
@@ -0,0 +1,198 @@
+  tar -xzf "${INFRA_DIR}/fs.tgz" -C "${FS_SYNC_DIR}"
+fi
+
+# Establish the server information
```
I would expect the configuration variables (including TASK_TIME_LIMIT and the three variables here) to be in their own section at the top of the file.

mernst commented on client_side/.infrastructure/setup.sh:
```diff
@@ -0,0 +1,198 @@
+FS_DIR="${EXP_DIR}/file_system"
+
+# The directory used by the infrastructure to reset FS_DIR.
+# It is created by extracting the fs.tgz tarball in INFRA_DIR.
```
Why extract it?  Why not just include it verbatim in the zip file?

mernst commented on client_side/.infrastructure/setup.sh:
```diff
@@ -0,0 +1,198 @@
+# It is created by extracting the fs.tgz tarball in INFRA_DIR.
+
+# The reason this isn't also distributed with the client ZIP is to prevent any
+# confusion as well as having a relatively smaller ZIP file.
```
This is likely to *increase* the size of the zip file.  A zip file uses compression.  If you do tar compression first, it has less data to work with and cannot compress as effectively.  Then, the zip compression can't do anything with the compressed tar file, so it also has less data to work with.

Using a tgz file seems to increase complexity without concomitant benefits.

mernst commented on client_side/.infrastructure/setup.sh:
```diff
@@ -0,0 +1,198 @@
+
+################################################################################
+#                              VARIABLE DEFINITIONS                            #
+#             This includes bash variables as well as variable files           #
```
I don't see variable files in this part of the script.

mernst commented on client_side/.infrastructure/setup.sh:
```diff
@@ -0,0 +1,198 @@
+  task_num=0
+fi
+
+# The TASK_ORDER is two two-character codes.  In each two-character code, the
```
The next two variables are constants that I would have expected to see set earlier in this file.

mernst committed:
More code review changes

mernst commented on client_side/.infrastructure/setup.sh:
```diff
@@ -0,0 +1,205 @@
+# Saves the command that was just entered by the user (and is about to be
+# executed) into the .command file.
+#
+# If the user enters an empty command, then the .command file does not change.
```
Why doesn't it change?  An explanation would be helpful.

mernst committed:
More code review changes

mernst commented on client_side/.infrastructure/infrastructure.sh:
```diff
@@ -0,0 +1,353 @@
+  echo "Welcome to the bash user study!"
+  echo "At any point, run \"helpme\" to see a list of commands available to you."
+
+  echo "You will have 5 minutes to complete each task."
```
This is premature at this point.  It should be printed only after training, when the user is getting started on actual tasks.

mernst committed:
More code review changes

mernst commented on client_side/.infrastructure/infrastructure.sh:
```diff
@@ -0,0 +1,358 @@
+      # If the user successfully did the infrastructure training, disables it.
+      unset INF_TRAINING
+
+      if [[ "${TEL_TRAINING:-false}" == "true" ]]; then
```
I suggest you delete this, and change the next `elif` to `fi; if` (with newline instead of `;`).  That avoids code duplication and ensures that this handles both types of training (if necessary), one after the other.

anhnamtran replied:
The reason for the `elif` is so that infrastructure training does not happen at
the same time as Tellina training. If we change the `elif` into an `if`, the
next if will still have to check whether we are in `INF_TRAINING`. This does
reduce the duplication by a little bit, but adds another level for the nested `if`s.

mernst replied:
I think the problem here is that I don't fully understand how training works.  I think this code doesn't do the training, just sets things up so that later code will do the training.  Maybe a brief explanation of that would have headed off my confusion.

anhnamtran replied:
I will add the explanation as part of the comment for this function.

mernst committed:
More code review changes

mernst commented on client_side/.infrastructure/infrastructure.sh:
```diff
@@ -0,0 +1,351 @@
+# pairs.
+tellina_training() {
+  echo "--------------------------------------------------------------------------------"
+  echo "This half of the experiment permits you to use the Tellina tool."
```
Why doesn't this use print_treatment instead of this line?  Either will work, I guess.

anhnamtran replied:
Print treatment happens right before either `infra_training` or `tellina_training`.

mernst commented on client_side/.infrastructure/infrastructure.sh:
```diff
@@ -0,0 +1,351 @@
+      reset_fs
+      ;&
+    *)
+      echo "Actual output does not match expected. A diff has been shown."
```
This might be a message about the file system state.  I feel it's important to have different messages for wrong file system state and wrong output.

mernst replied:
This does not seem to have been resolved.  I expected to see
```
case $exit_code in
0)
...
1)
...
2)
...
3)
...
```

but the same message is still being printed for cases 1 and 3.
I think the message about "actual output" will be confusing if the user just issued a command that produced no output (the command only changed the file system).

anhnamtran replied:
Ah I misunderstood what you meant.

mernst committed:
Improve comments

mernst commented on client_side/.infrastructure/infrastructure.sh:
```diff
@@ -0,0 +1,351 @@
+
+# Prints the list of resources that the user is allowed to use based on the
+# current treatment.
+print_treatment() {
```
Should helpme also call print_treatment?

mernst reviewed (changes requested):
Thanks!  This is looking much better.
I have some suggestions and comments.

mernst committed:
Clarify message

anhnamtran committed:
Use raw file system directory instead of extracting it in setup

anhnamtran committed:
Addressed code review comments about variables and comments

anhnamtran committed:
Changed "abandon" -> "giveup"

anhnamtran committed:
Reduced some code duplication

anhnamtran committed:
Changed some prompts

anhnamtran committed:
Revert "Reduced some code duplication"

This reverts commit c5c6f0a48ec41920341bb23f83b0113d365e5ad9.

anhnamtran committed:
Added explanation for check priority in check_and_update_training_status

anhnamtran committed:
Moved a prompt to be less premature

anhnamtran committed:
More code review changes

anhnamtran committed:
Make sure that modification times and permisssions are correct for client's file_system directory

anhnamtran committed:
Track empty file_system directories

anhnamtran committed:
Remove no longer needed tar ball

anhnamtran committed:
Make sure that we don't delete the gitkeep files

anhnamtran committed:
Fix incorrect regex

anhnamtran commented:
@mernst 
I've address most of the code review comments and added more to the Makefile. Please take a final look. Once you've approved, please go ahead and merge or let me know so I can merge!

anhnamtran committed:
Correct quoting for paths

anhnamtran committed:
Re-run travis build

anhnamtran committed:
Removed unused line of code

anhnamtran commented:
The build is failing in Travis but not locally. I am unsure why and I'm investigating it right now. 

Edit: This is because of the `.gitkeep` files for the empty directories. This is fine, since I can just remove it during configuration.

anhnamtran committed:
Some debug information in tests

anhnamtran committed:
Use global setup file instead

anhnamtran committed:
Preserve some properties of files in reset_fs, exclude hidden files

anhnamtran commented:
I realized that the `rsync` command run by `reset_fs` also did not preserve modification times or permissions, so I've also addressed that issue.

anhnamtran committed:
Only preserve modification time

anhnamtran committed:
[skip travis] Use variable for most of link instead of hard-coded

anhnamtran committed:
Adjusted design document

mernst committed:
Tweaks

mernst committed:
Tweaks

mernst committed:
Move variables around

mernst committed:
Change "abandon" to "giveup"

mernst committed:
Add punctuation

mernst committed:
Undo accidental change to binary

mernst committed:
Move variables

mernst commented on client_side/.infrastructure/verify_task.py:
```diff
@@ -0,0 +1,188 @@
+        print(e)
+        sys.exit(SUBPROCESS_FAILURE)
+
+def normalize_and_copy_output(out_file, norm_file):
```
I find the formal parameter `out_file` very confusing.  It is an input to this routine.  If anything were named `out_file`, it should be `norm_file`.  I see why you name it this way -- because clients of this routine will pass a file that is the output of some other command.  It's still confusing here.  You should write the specification of a routine from that routine's point of view, not with respect to some particular client.

mernst reviewed (changes requested):
Andrew, I have just two more things that I think should be modified.

mernst commented:
Note that one of my recent two comments is marked resolved, but it doesn't look to me like it is resolved.  (Am I wrong about that?)

anhnamtran commented:
I've unresolved it until I get to it

anhnamtran committed:
Split up case statement for verify task

anhnamtran committed:
Changed some parameter names

anhnamtran committed:
Changed a timing test to be less harsh

mernst committed:
More terse messages

mernst reviewed (approved):
This looks great.
You can squash-and-merge it whenever you like, then move the repository to the TellinaTool organization.

Thanks very much for your hard work on this, which has yielded a comprehensive and comprehensible solution.
